### ISSUE TITLE
changefeedccl: Incorrect retry logic in webhook sink

### ISSUE THREAD

**Issue by miretskiy** [cockroachdb/cockroach#118485]
Webhook sink retries are broken.  Any error returned by downstream
sink (for example: overloaded, go away, etc), when retried, would re-send 
a request that will have a non-zero content length, but zero length body.

This is due to the fact that the constructed http request:
`req, err := http.NewRequestWithContext(sc.ctx, http.MethodPost, sc.url.String(), bytes.NewReader(body))`
initializes request "GetBody" callback which is called by http2 library whenever
it needs to retry the request.  However, internally, the library retries very limited
number of errors (such as http redirects) and all other errors are propagated to the caller.
However, when we issue our own internal retry 
```
return retry.WithMaxAttempts(ctx, retryOpts, retryOpts.MaxRetries+1, func() error {
		return sc.Flush(ctx, pl)
	})
```
the http request now has "Body" initialized, and that reader has already consumed
all of the data.

Basically, we need to re-initialize req.Body field every time we retry.
See https://github.com/cockroachdb/cockroach/pull/118480

The above PR needs to be augmented with a unit test: something that spins
up local http end point; ensure an error is returned; then ensure that upon retry,
we do not send a non-zero content-length with zero length body.

Jira issue: CRDB-35757

**Comment by bobvawter** [comment]
Any HTTP/2 webhook sink that received one of these possibly-shortened, retried payloads and which did not verify the `Content-Length` versus actual bytes received or those which don't parse the payload during receipt may have lost data.

**Comment by bobvawter** [comment]
For archival purposes, here's a non-CRDB, minimum repro that was useful while investigating the behavior of the go HTTP/2 behaviors: https://gist.github.com/bobvawter/c5fb99bbc8de533bb53881fe78b8db44

It's interesting to note that running in in HTTP/1.1 mode did not seem to reliably trigger the short-write behavior that we saw in HTTP/2 mode.

**Comment by glennfawcett** [comment]
Running tests on v23.2.2 and still seeing this bug...

```
	time="Mar 13 20:21:32" level=error detail="request declared a Content-Length of 1050776 but only wrote 0 bytes\ncould not decode payload\ngithub.com/cockroachdb/cdc-sink/internal/source/cdc.(*Handler).webhook\n\t/home/runner/work/cdc-sink/cdc-sink/internal/source/cdc/webhook.go:52\ngithub.com/cockroachdb/cdc-sink/internal/source/cdc.(*Handler).ServeHTTP\n\t/home/runner/work/cdc-sink/cdc-sink/internal/source/cdc/handler.go:104\ngithub.com/cockroachdb/cdc-sink/internal/util/stdserver.logWrapper.func1\n\t/home/runner/work/cdc-sink/cdc-sink/internal/util/stdserver/wrapper.go:105\nnet/http.HandlerFunc.ServeHTTP\n\t/opt/hostedtoolcache/go/1.20.14/x64/src/net/http/server.go:2122\nnet/http.(*ServeMux).ServeHTTP\n\t/opt/hostedtoolcache/go/1.20.14/x64/src/net/http/server.go:2500\ngolang.org/x/net/http2/h2c.h2cHandler.ServeHTTP\n\t/home/runner/go/pkg/mod/golang.org/x/net@v0.21.0/http2/h2c/h2c.go:125\nnet/http.serverHandler.ServeHTTP\n\t/opt/hostedtoolcache/go/1.20.14/x64/src/net/http/server.go:2936\nnet/http.initALPNRequest.ServeHTTP\n\t/opt/hostedtoolcache/go/1.20.14/x64/src/net/http/server.go:3545\nnet/http.(*http2serverConn).runHandler\n\t/opt/hostedtoolcache/go/1.20.14/x64/src/net/http/h2_bundle.go:6111\nruntime.goexit\n\t/opt/hostedtoolcache/go/1.20.14/x64/src/runtime/asm_amd64.s:1598" error="could not decode payload: request declared a Content-Length of 1050776 but only wrote 0 bytes" uri=/rodan/public
	```

**Comment by rharding6373** [comment]
As discussed in https://cockroachlabs.slack.com/archives/C9TGBJB44/p1710448711619219, the backport was not in 23.2.2.
